content/blog/2022-05-17-object-oriented-features/index.md

```diff
@@ -0,0 +1,210 @@
++++
+title = "Xi++: Object Oriented Features for Xi"
+[[extra.authors]]
+name = "Orko Sinha"
+link = ""
+[[extra.authors]]
+name = "Michael Maitland"
+link = "https://michaelmaitland.com"
++++
+Building a compiler for Xi is the main project in [CS 4120](https://www.cs.cornell.edu/courses/cs4120/) at Cornell. The final assignment, now discontinued, is to extend the compiler to support feaures of the [Xi++ Language Specification](https://www.cs.cornell.edu/courses/cs4120/2019sp/project/oolang.pdf) such as object-oriented programing. Our project is an implementation of these features on Michael's compiler created for CS 4120. Our project is [here](https://github.com/Yasgur99/xic/tree/oo).
+
+# Lexing
+For lexing, there are 5 new syntax elements that we need to handle, listed below:
+1. `class`
+2. `extends`
+3. `this`
+4. `new`
+5. `.` operator
+
+These correspond to 5 new token types corresponding to each keyword in the language. The existing JFlex file created for the original project is extended to support these features.
+
+# Parsing
+The parser handles the new semantics and allow parsing of the following:
+1. Parsing of `Interface` that contain zero or more `use` statements and zero or more class declarations. A class declaration is the name of the class, optionally the class name it extends, and zero or more function declarations. `ClassDecl` is created as a new AST node to represent this.
+2. Parsing of `Program` that contain class definitions and function definitions. Programs can now contain one or more functions definitions and/or class definitions. A class definition is the name of the class, and the class it extends a list of fields which are `SimpleDecl`s, and a list of function defintions. However, the list of fields or methods can be empty. `ClassDefn` also represents this.
+3. Parsing of `Exp` with the `new` and `this` keyword.
+4. Parsing of `MethodCall`s which take the form `o.f()` or `this.f()`
+5. Parsing of `FieldAccess` which take the form of `o.x` or `this.x`
+
+Additionally, two AST nodes are added to handle Interfaces and Classes.
+1. `InterfaceBody`
+2. `ClassBody`
+
+Previously, an `Interface` and `Program` contained the function declarations and definitions respectively. Previously, the parser would add to the lists directly when it came across a declaration or definition. Now that we are also handling class declarations and definitions, the parser has `fdecl_or_cdecl_star` and `fdefn_or_cdefn_plus` nonterminal to account for this. These two new nonterminals correspond to the two new AST nodes. This does not have a big input on the rest of the compiler as we just need to add a `getBody()` call in certain places that expected the function declarations and definitions to be accessed directly from the `Program`.
+
+We also add AST nodes `FieldAccess`, `MethodCall`, and `New`.
```

Comment:
Again, I don't think the actual class names are at the right level of abstraction for readers of this blog post.